### ISSUE TITLE
Neolink fails to connect to devices not using encryption

### ISSUE THREAD

**Issue by twistedddx** [thirtythreeforty/neolink#79]
-**Describe the bug**
Neolink fails to log directly in to a Swann NHD-818.
When connecting via NVR running Reolink firmware, eg NVR is a proxy to the camera, Neolink can connect correctly.
Swann NHD-818 uses modern messages but does not encrypt the XML, it is clear text.

Observing packet capture between Reolink client and camera the rest of the communication will be in clear text without encryption and encrypted flag set to false.

Neolink sends legacy login. Header encrypted flag set to true but Reolink client does the same.
Camera replies with clear text encryption XML with md5 nonce. Header encrypted flag is false.
Neolink no longer processes connection and instead retries and returns to legacy login step.

Neolink log shows:
C:\gstreamer\1.0\x86_64\bin>C:\gstreamer\1.0\x86_64\bin\neolink.exe --config neolink_config.toml
[2020-09-26T01:52:05Z INFO neolink] Neolink v0.3.0-123-gaffc0d9 release
[2020-09-26T01:52:05Z DEBUG neolink::gst] Permitting anonymous to access /1, /1/mainStream
[2020-09-26T01:52:05Z DEBUG neolink::gst] Permitting anonymous to access /1/subStream
[2020-09-26T01:52:05Z INFO neolink] 1: Connecting to camera at 192.168.93.44:9000
[2020-09-26T01:52:05Z INFO neolink] 1: Connecting to camera at 192.168.93.44:9000
[2020-09-26T01:52:05Z DEBUG neolink::bc_protocol::connection] Shutting down BcConnection...
[2020-09-26T01:52:05Z DEBUG neolink::bc_protocol::connection] Shutdown finished OK
[2020-09-26T01:52:05Z ERROR neolink] Error streaming from camera 1, will retry in 1s: Timeout

**To Reproduce**
Steps to reproduce the behavior. Example:
1. Create this configuration file:
bind = "0.0.0.0"

[[cameras]]
name = "1"
username = "admin"
password = "12345"
address = "172.16.25.2:9000"
format = "h264"

2. Launch Neolink:
3. Watch failure

**Expected behavior**
Neolink to connect

**Versions**
NVR software: 541_17051800
Neolink software: v0.3.0-123-gaffc0d9
Reolink camera model and firmware: Swann NHD-818 aka IPC_388M

**Comment by twistedddx** [comment]
I think the issue is in ser.rs and the bc_xml function being used to encrypt/decrypt xml regardless of self.meta.encrypted

**Comment by QuantumEntangledAndy** [comment]
Wow unencrypted (well unobsfucated more like), I didnt know we weren't respecting the encrypted flag in the packets but if that's the case this should be a relatively simple fix. Just need an if block around the encrypt/decrypt. We may need to add if the camera replied unencrypted into the context though.

**Comment by QuantumEntangledAndy** [comment]
Should we be supporting unencrypted communication? Isn't it good behaviour to refuse to send password unencrypted?

I suppose it's a limit of the camera, but sigh.... BC....

**Comment by QuantumEntangledAndy** [comment]
If we were to do this then we should change 

```suggestion

let modern_login = Bc::new_from_xml(
            BcMeta {
                msg_id: MSG_ID_LOGIN,
                client_idx: 0, // TODO
                encrypted: true,
                class: 0x6414,
            },

```

In bc_protocol and add an encryption flag to Bc_Camera that's set based on the reply to the login and is then used in all future messages.

**Comment by twistedddx** [comment]
The username and password of all BC devices is exposed in the legacy login step.
The "upgrade" to modern with a nonce and then upgraded to encrypted XML all appear to be just extra layers of weak obsfucation.

Swann/BC/Reolink design for these cameras was to direct connect to NVR.
To attack this clear text in Swann/BC/Reolink design you would need to compromise the NVR, camera or physical cable.
Neolink is interupting that design and perhaps increasing risk by inviting traffic to cross other devices.

I vote note the weak security of username/passwords in BC devices in the readme and allow the Neolink user to decide if it concerns them.

**Comment by QuantumEntangledAndy** [comment]
For some reason I thought we were sending a dummy login during legacy just to get the noonce but it seems I misread the code. But yes I think we can implement this as you said and have a waning in the log too

**Comment by QuantumEntangledAndy** [comment]
I shall add it to my to-do list. Seems my list of PR will be using up lots of thirtythreeforty time though. Sorry if it's too much :(. Maybe should add this one once everything else is merged.

**Comment by twistedddx** [comment]
What I see is the official client(Win x64 7.2.2.33) always keeps its encrypted flag false after the legacy login which had it true.
Also NHD-818 changes to saying it is encrypted when it clearly isn't.
Either the field isn't what I think it is or simply can not be trusted other than from the camera itself the first time it responds.

Eg
Reolink RLC-410 which does use encryption:
Client sends cleartext legacy login with encrypted true.
Cam responds with encrypted modern upgrade request with encrypted true  << correct
Client sends encrypted login with encrypted false
Cam responds with encrypted DeviceInfo with encrypted true
Client sends encrypted AbilityInfo with encrypted false
Cam responds with encrypted AbilityInfo with encrypted true

Swann NHD-818 which does not use encryption:
Client sends cleartext legacy login with encrypted true.
Cam responds with cleartext modern upgrade request with encrypted false  << correct
Client sends cleartext login with encrypted false
Cam responds with cleartext DeviceInfo with encrypted true
Client sends cleartext AbilityInfo with encrypted false
Cam responds with cleartext AbilityInfo with encrypted true

**Comment by QuantumEntangledAndy** [comment]
It seems we need to cache the encryptable status at the modern upgrade step then use that throughout. That's not too difficult to do just lots of changes everywhere we send messages to used the cached value.

I also haven't figured out if our serialiser respects what the flag is set to. The seralise XML part seems to be nested inside a match and I havent unwrapped its meaning yet. Seems my rust foo needs an upgrade :)... Or I'm just looking in the wrong place haha.

**Comment by twistedddx** [comment]
Fixed in #83
